Update Bash highlights

David-Else opened the pull request:
I have updated the bash highlights and tree sitter version, it is much better now. The highlights are based on Neovim's latest https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/bash and converted for Helix. I removed the `#lua-match?` lines, I am afraid I don't know if there is a Helix equivalent?

I tested it by loading Helix and Neovim in a split window with the equivalent VS Code theme, and they now look almost identical. The only flaw I can find is that the comments in Helix highlight anything starting with an `@`, how can I fix it?

### Helix left, Neovim right
![image](https://github.com/user-attachments/assets/7620ba92-d8fe-4850-aba3-6f8de2b0e33c)
![image](https://github.com/user-attachments/assets/d57205bf-f687-40db-8f4b-39daf8d5ca47)


I updated the grammar to the latest [v0.23.3](https://github.com/tree-sitter/tree-sitter-bash/releases/tag/v0.23.3), but I can't seem to get Helix to update to it and build! I don't think it needs the new grammar as there have not been many changes, but I thought it would be good form to have the latest. I have tried deleting the source and compiled version and `hx --grammar fetch` and build but it keeps on using the old version, is it stuck in some cache? I will mark it draft until I can confirm all is well with that. I asked again for help here: https://github.com/helix-editor/helix/discussions/13482

I would like to get the heredoc injection working, but I have no idea how, here is the Neovim version:

```scm
((heredoc_redirect
  (heredoc_body) @injection.content
  (heredoc_end) @injection.language)
  (#downcase! @injection.language))
```

To make it easy to see what I have done, here is a diff with the Neovim version:

```diff
--- /home/david/Downloads/neovim-bash-highlights.scm	2025-05-06 12:34:16.710777764 +0100
+++ /home/david/github/David-Else/helix/runtime/queries/bash/highlights.scm	2025-05-06 20:40:36.708969117 +0100
@@ -46,9 +46,6 @@
   "!"
 ] @operator
 
-; Do *not* spell check strings since they typically have some sort of
-; interpolation in them, or, are typically used for things like filenames, URLs,
-; flags and file content.
 [
   (string)
   (raw_string)
@@ -79,7 +76,7 @@
   "case"
   "in"
   "esac"
-] @keyword.conditional
+] @keyword.control.conditional
 
 [
   "for"
@@ -88,7 +85,7 @@
   "select"
   "until"
   "while"
-] @keyword.repeat
+] @keyword.control.repeat
 
 [
   "declare"
@@ -99,7 +96,7 @@
   "unsetenv"
 ] @keyword
 
-"export" @keyword.import
+"export" @keyword.control.import
 
 "function" @keyword.function
 
@@ -117,10 +114,10 @@
     "SIGRTMAX-12" "SIGRTMAX-11" "SIGRTMAX-10" "SIGRTMAX-9" "SIGRTMAX-8" "SIGRTMAX-7" "SIGRTMAX-6"
     "SIGRTMAX-5" "SIGRTMAX-4" "SIGRTMAX-3" "SIGRTMAX-2" "SIGRTMAX-1" "SIGRTMAX"))
 
-((word) @boolean
-  (#any-of? @boolean "true" "false"))
+((word) @constant.builtin.boolean
+  (#any-of? @constant.builtin.boolean "true" "false"))
 
-(comment) @comment @spell
+(comment) @comment
 
 (test_operator) @operator
 
@@ -149,7 +146,7 @@
   [
     "?"
     ":"
-  ] @keyword.conditional.ternary)
+  ] @keyword.control.conditional)
 
 (binary_expression
   operator: _ @operator)
@@ -164,7 +161,7 @@
   name: (word) @function)
 
 (command_name
-  (word) @function.call)
+  (word) @function)
 
 (command_name
   (word) @function.builtin
@@ -188,10 +185,7 @@
 (unset_command
   (word) @variable.parameter)
 
-(number) @number
-
-((word) @number
-  (#lua-match? @number "^[0-9]+$"))
+(number) @constant.numeric
 
 (file_redirect
   (word) @string.special.path)
@@ -214,19 +208,19 @@
 (expansion
   "@"
   .
-  operator: _ @character.special)
+  operator: _ @constant.character)
 
 ((expansion
   (subscript
-    index: (word) @character.special))
-  (#any-of? @character.special "@" "*"))
+    index: (word) @constant.character))
+  (#any-of? @constant.character "@" "*"))
 
 "``" @punctuation.special
 
 (variable_name) @variable
 
 ((variable_name) @constant
-  (#lua-match? @constant "^[A-Z][A-Z_0-9]*$"))
+  (#match? @constant "^[A-Z][A-Z_0-9]*$"))
 
 ((variable_name) @variable.builtin
   (#any-of? @variable.builtin
@@ -254,8 +248,3 @@
   (regex)
   (extglob_pattern)
 ] @string.regexp
-
-((program
-  .
-  (comment) @keyword.directive @nospell)
-  (#lua-match? @keyword.directive "^#!/"))
```

nik-rev commented:
Helix has `#match?` as the analogue to `#lua-match?`

```scm
((identifier) @type
  (#match? @type "^[A-Z][a-z0-9_]+"))
```

David-Else commented:
@nik-rev Thanks for the tip!

I fixed the constant variables with:

```
((variable_name) @constant
  (#match? @constant "^[A-Z][A-Z_0-9]*$"))
```

but with:

```
(number) @constant.numeric

((word) @constant.numeric
  (#match? @constant.numeric "^[0-9]+$"))
```
The 2nd `#match? @constant.numeric "^[0-9]+$")` seems redundant? I still gets numbers correctly highlighted without it, am I missing something? Maybe it is just old code that I can delete? Cheers :)

nik-rev commented:
> @nik-rev Thanks for the tip!
> 
> I fixed the constant variables with:
> 
> ```
> ((variable_name) @constant
>   (#match? @constant "^[A-Z][A-Z_0-9]*$"))
> ```
> 
> but with:
> 
> ```
> (number) @constant.numeric
> 
> ((word) @constant.numeric
>   (#match? @constant.numeric "^[0-9]+$"))
> ```
> 
> The 2nd `#match? @constant.numeric "^[0-9]+$")` seems redundant? I still gets numbers correctly highlighted without it, am I missing something? Maybe it is just old code that I can delete? Cheers :)

I assume a previous version of tree-sitter-bash required these but it's no longer necessary. If there's no different, might be safe to remove it! Though I'd personally keep it there just in case, if it does not intervene with anything else

David-Else commented:
> > @nik-rev Thanks for the tip!
> > I fixed the constant variables with:
> > ```
> > ((variable_name) @constant
> >   (#match? @constant "^[A-Z][A-Z_0-9]*$"))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but with:
> > ```
> > (number) @constant.numeric
> > 
> > ((word) @constant.numeric
> >   (#match? @constant.numeric "^[0-9]+$"))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The 2nd `#match? @constant.numeric "^[0-9]+$")` seems redundant? I still gets numbers correctly highlighted without it, am I missing something? Maybe it is just old code that I can delete? Cheers :)
> 
> I assume a previous version of tree-sitter-bash required these but it's no longer necessary. If there's no different, might be safe to remove it! Though I'd personally keep it there just in case, if it does not intervene with anything else

I tested it and decided to delete it out of a hatred of bloat. I also pushed the fix you suggested for the constant highlighting and updated the diff in my original post.

David-Else commented:
[I could not seem to get bash-treesitter to use the updated revision in languages.toml](https://github.com/helix-editor/helix/discussions/13482), so I updated it manually before `hx --grammar build` to test this update:
![image](https://github.com/user-attachments/assets/4579318e-1c90-4514-a055-2a1b754536e6)
I can confirm there were no problems so this is now ready for review.

@the-mikedavis Is converting this Neovim code to Helix possible?

```
((heredoc_redirect
  (heredoc_body) @injection.content
  (heredoc_end) @injection.language)
  (#downcase! @injection.language))
```
Cheers.

the-mikedavis commented:
We don't currently support a `#downcase!` property but that's something we could add support for. For now that can probably just be commented out and we can revisit it later

David-Else commented:
> We don't currently support a `#downcase!` property but that's something we could add support for. For now that can probably just be commented out and we can revisit it later

I have not added any of the new injections from Neovim, I was just asking. The PR should be OK to merge once you have given it a check :)
